Document patterns and suggestions for semconv code generation

lmolkova opened the pull request:
Follow up on #551

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* ~~[schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.~~

jsuereth reviewed (commented):
This LGTM - Think we should merge and promote this.

I think it *may* be possible to document weaver here as well, but I'd prefer *one* recommended solution.  Would like to hear from @lquerel if we think weaver is ready for this level of codegen.  If weaver is our long-term play for codegen, would be nice to have it be the base recommendation when we reach out beyond java/python.

lquerel commented on supplementary-guidelines/semantic_conventions_code_generation.md:
```diff
@@ -0,0 +1,94 @@
+OpenTelemetry Language SIGs can auto-generate semantic conventions code in the form that's idiomatic for
+their language and may (or may not) ship semantic-conventions as a stand-alone library.
+
+This document outlines common patters and provides non-normative guidance on how to implement auto-generation
```
```suggestion
This document outlines common patterns and provides non-normative guidance on how to implement auto-generation
```

lquerel commented on supplementary-guidelines/semantic_conventions_code_generation.md:
```diff
@@ -0,0 +1,94 @@
+Other possible solutions include:
+
+- Generate all semantic conventions for a given version in specific folder while keeping old versions. It is used by [opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go/tree/main/semconv/) but could be problematic if the artifact size is a concern.
+- Follow language-specific conventions to annotate experimental parts. For example, Semantic Conventions in Python keeps experimental attributes in `opentelemetry.semconv._incubating` import path which is considered (following Python underscore convention) to be internal and subject to change.
```
Other languages like Rust supports the concept of `feature` for this kind of optional/experimental code.

lquerel commented on supplementary-guidelines/semantic_conventions_code_generation.md:
```diff
@@ -0,0 +1,94 @@
+### Recommendations
+
+- Generate code for deprecated attributes, metrics, and other conventions (even if they never reached stability). It minimizes runtime issues and breaking changes in user applications. Use appropriate annotations to mark deprecated conventions. All deprecated conventions should have `deprecated` property describing new alternatives (when available) which can be used in auto-generated code documen ...
+- Generate code for stable convention inside incubating artifact. It prevents user code from breaking when semantic convention stabilizes. Deprecate stable definitions inside incubating artifact and point users to the stable location.
```
Upon first reading, I did not understand this point. The confusion probably stems from the fact that we only discuss `stable` or `deprecated` semconv, but not semconv marked as `experimental`. I think we could gain clarity by specifying that the purpose of this artifact is to generate `experimental` semconv in addition to the other two levels of stability.

trask replied:
Does the confusion come from experimental == incubating?

Would it be better to use the same term (experimental/incubating) everywhere instead of two different terms?

cc @jack-berg

lmolkova replied:
I've rephased it and moved this section to be under the artifact structure. I hope it's more clear.

I'm also very curious on everyone's thoughts regarding experimental/incubating/etc.

lquerel commented:
@lmolkova and @jsuereth I think Weaver can certainly be used for code generation (this was in fact the first use). I believe that the solution currently proposed in Weaver is more flexible than that supported by a raw Ninja version. 

For example, in Weaver, it is possible to generate multiple code files from the same template. It is also possible to define the name of the generated file programmatically. To make life a little easier for those in charge of generating code for semconv in different programming languages, we could go even further by adding some Jinja filters and tests like `to_doc_brief` and others to be defined based on existing templates.

lmolkova commented on supplementary-guidelines/semantic_conventions_code_generation.md:
```diff
@@ -0,0 +1,94 @@
+> Shipping two versions of the same artifact (stable and preview) could be problematic if dependency resolution mechanism is based on the latest version.
+> For example, if user application depends on the `semconv v1.0.0-preview` and some library brings transitive dependency on `semconv v1.1.0`, the latter would be resolved leading
+> to compile or runtime issues in the application.
+> Shipping two artifacts with different names is recommended for the ecosystems that are prone to diamond-dependency conflicts.
```
reorganized  the text a bit to be more clear

lmolkova commented:
@lquerel @jsuereth 

My main goal with this is to document what code generation should produce in terms of stability, structure, naming, etc which should not depend on the tooling.

I've moved the tooling section to be the last and added this note:

> Note:
> The tooling used for code generation may change to [opentelemetry-weaver](https://github.com/open-telemetry/weaver),
> without any breaking changes in the generated code and with minimal changes to generation process and templates.

WDYT?

lquerel commented:
@lmolkova 

> [Update] we should definitely look into moving codegen to weaver, would be great to have a example similar to #917 but with code.

Which repo will be a good example for codegen ?

lmolkova commented:
@lquerel 

> Which repo will be a good example for codegen ?

[Java](https://github.com/open-telemetry/semantic-conventions-java/blob/main/buildscripts/templates/SemanticAttributes.java.j2) one is the most advanced one that follows the guidance in this doc.

> UPDATE: I can create a PR for the Rust client SDK semconv section.

 That would be awesome 🙌

[Rust](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-semantic-conventions/scripts/templates/semantic_attributes.rs.j2) is following old undocumented patterns :)

lquerel commented:
> [Java](https://github.com/open-telemetry/semantic-conventions-java/blob/main/buildscripts/templates/SemanticAttributes.java.j2) one is the most advanced one that follows the guidance in this doc.
> 
> 
> 
> > UPDATE: I can create a PR for the Rust client SDK semconv section.
> 
> 
> 
>  That would be awesome 🙌
> 
> 
> 
> [Rust](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-semantic-conventions/scripts/templates/semantic_attributes.rs.j2) is following old undocumented patterns :)
> 

Okay, I will review the Java version and use this experiment as an opportunity to propose a more up-to-date version of the Rust implementation.

joaopgrassi commented on supplementary-guidelines/semantic_conventions_code_generation.md:
```diff
@@ -0,0 +1,118 @@
+> experimental convention, the latter would be resolved leading to compilation or runtime issues in the application.
+
+Instrumentation libraries should depend on the stable (parts of) semantic convention artifact or copy relevant definitions into their own code base.
+Incubating parts are intended for the end-users.
```
What does "Incubating parts" mean here?

joaopgrassi replied:
We talk about Incubating in several places it seems it's never defined what it means. above we kinda say it's where experimental attributes should go in, but here we say they are intended for end user. Below we say it's for stable conventions. I'm  a bit confused 🤔

lmolkova replied:
I've updated the text to avoid using it except in the naming suggestions.

It effectively means either a package like `semconv-incubating` that contains all the semantic conventions (experimental and stable). Or, if there is just one package, it's a part that's annotated as not stable. E.g. in python it's `semconv._incubating`.

I now use `preview` to refer to that artifact, lmk if it's better or if there are parts I still need to clarify. Thanks!
